From dcd81d7a8cff4b7be5e6eb9482817e5938010148 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:06 -0400
Subject: [block] blk-mq: allow setting of per-request timeouts

Message-id: <1402670238-13916-79-git-send-email-snitzer@redhat.com>
Patchwork-id: 83828
O-Subject: [RHEL7.1 PATCH 078/150] blk-mq: allow setting of per-request timeouts
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit c22d9d8a60646a1521ae12a2176f58da19afa186
Author: Jens Axboe <axboe@fb.com>
Date:   Fri May 23 14:14:57 2014 -0600

    blk-mq: allow setting of per-request timeouts

    Currently blk-mq uses the queue timeout for all requests. But
    for some commands, drivers may want to set a specific timeout
    for special requests. Allow this to be passed in through
    request->timeout, and use it if set.

    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 54e7886..9bc0753 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -454,9 +454,13 @@ static void blk_mq_start_request(struct request *rq, bool last)
  /*
   * Just mark start time and set the started bit. Due to memory
   * ordering, we know we'll see the correct deadline as long as
-  * REQ_ATOMIC_STARTED is seen.
+  * REQ_ATOMIC_STARTED is seen. Use the default queue timeout,
+  * unless one has been set in the request.
   */
- rq->deadline = jiffies + q->rq_timeout;
+ if (!rq->timeout)
+  rq->deadline = jiffies + q->rq_timeout;
+ else
+  rq->deadline = jiffies + rq->timeout;
 
  /*
   * Mark us as started and clear complete. Complete might have been
-- 
1.7.1